app/controllers/projects_controller.rb

```diff
@@ -51,39 +51,128 @@ def index
                    else
-                     current_power.samples
+                     current_power.projects
                    end
 
-        @projects = if only_library || exclude_library
-                      @samples.group(:project).count
-                    else
-                      # Make sure you still return projects without any samples. Ex: Project listing
-                      # used by the CLI.
-                      (only_updatable ? current_power.updatable_projects : current_power.projects).map do |p|
-                        [p, Sample.where(project: p).count]
-                      end
-                    end
-        extended_projects = @projects.map do |project, sample_count|
+        sample_count_by_project_id = Hash[samples.group_by(&:project_id).map { |k, v| [k, v.count] }]
+        host_genome_names_by_project_id = {}
+        tissues_by_project_id = {}
+        samples.includes(:host_genome).each do |s|
+          (host_genome_names_by_project_id[s.project_id] ||= Set.new) << s.host_genome.name if s.host_genome.name
+          # TODO: sample_tissue column is deprecated, retrieve sample_type from Metadatum model instead
+          (tissues_by_project_id[s.project_id] ||= Set.new) << s.sample_tissue if s.sample_tissue
+        end
+        extended_projects = projects.map do |project|
           project.as_json(only: [:id, :name, :created_at, :public_access]).merge(
-            number_of_samples: sample_count,
-            hosts: @samples.where(project_id: project.id).includes(:host_genome).distinct.pluck("host_genomes.name").compact,
-            tissues: @samples.where(project_id: project.id).distinct.pluck(:sample_tissue).compact
+            number_of_samples: sample_count_by_project_id[project.id] || 0,
+            hosts: host_genome_names_by_project_id[project.id] || [],
+            tissues: tissues_by_project_id[project.id] || []
           )
         end
         render json: extended_projects
       end
     end
   end
 
+  def dimensions
+    # TODO(tiago): consider split into specific controllers / models
+    domain = params[:domain]
+
+    samples = samples_by_domain(domain)
+    # Filter by access control
+    sample_ids = samples.pluck(:id)
+
+    project_ids = samples.distinct(:project_id).pluck(:project_id)
+    projects = Project.where(id: project_ids)
+
+    # locations
+    locations = Metadatum
+                .joins(
+                  :metadata_field,
+                  :sample
+                )
+                .where(
+                  metadata_fields: { name: "collection_location" },
+                  sample_id: sample_ids
+                )
+                .group(:string_validated_value)
+                .distinct
+                .count(:project_id)
+
+    # for metadata fields we need to send both value and text
+    locations = locations.map do |location, count|
+      { value: location, text: location, count: count }
+    end
+
+    tissues = Metadatum
```

Comment:
I'd like to see all the types of metadata queries split into their own private methods because they are logically independent. Then we could maybe run them in parallel. Does rails support that?

Reply:
The problem here is redundancy since parallelization is orthogonal. Apparently there is promise gem we can explore. 

I will move the method to the model.